dispatch/static/manager/src/js/vendor/dispatch-editor/embeds/TwitterEmbed.js

```diff
@@ -0,0 +1,75 @@
+import React from 'react'
+
+import fetchJsonp from 'fetch-jsonp'
+import url from 'url'
+
+import { Button } from '@blueprintjs/core'
+import { FormInput, TextInput } from '../../../components/inputs'
+
+require('isomorphic-fetch')
+require('../styles/embeds/video.scss')
+
+const TWITTER_API_URL = 'https://api.twitter.com/1.1/statuses/oembed.json'
+
+class TwitterEmbedComponent extends React.Component {
+
+  validateTweetURL(url) {
+    return /^(?:https?:\/\/)?(?:www\.)?twitter\.com\/?(\w+)\/(status)\/(\d+)/g.test(url)
+  }
+
+  getTweet() {
+    const that = this
+    if (this.validateTweetURL(this.props.data.url)) {
+      fetchJsonp(TWITTER_API_URL + url.format({query: {url: this.props.data.url}}))
+      .then((response) => response.json())
+      .then(data => {
+        that.props.updateField('tweet', data.html)
+        this.props.stopEditing()
+      })
+    }
+  }
+
+  renderInput() {
+    return (
+      <div className='o-embed o-embed--tweet'>
+        <form>
+          <FormInput label='URL'>
+            <TextInput
+              fill={true}
+              value={this.props.data.url}
+              onChange={e => this.props.updateField('url', e.target.value)} />
+          </FormInput>
+        </form>
+        <div className='o-embed--video__button__container'>
+          <Button onClick={() => this.getTweet()}>Insert</Button>
+        </div>
+      </div>
+    )
+  }
+
+  renderTweet() {
+    return (
+      <div>
+        <div dangerouslySetInnerHTML={{__html: `${this.props.data.tweet}`}} />
```

Comment:
Don't know if this has been talked about in person at meetings or something, but I'd advise against using `dangerouslySetInnerHTML` as it does open you up to XSS attacks. This is especially dangerous when the data source is a third party (ie. twitter), and out of your control.

If you insist on using `dangerouslySetInnerHTML`, then you should first sanitize the html before injecting it.

Reply:
This was indeed discussed in person, but since I don't have too much experience with Javascript, HTML etc., I'm more than open to suggestions on how to replace that dangerouslySetInnerHTML!

Reply:
So for context, the danger here is there could be `<script>` tags in the tweet, which would allow a malicious actor to run any JS they want when we embed the tweet. I would just sanitize the html returned from the twitter api first before injecting it, its the simplest solution. 

The more involved solution would be to perform the tweet fetching on the server side and perform the sanitation there before passing clean html to the client. This would make it so users can't bypass the sanitation, which I suppose isn't really your responsibility; if they decide to bypass the sanitation, then they are responsible for the outcome.